.github/workflows/default.yaml

```diff
@@ -122,7 +129,11 @@ jobs:
 
       - name: Run build on Linux
         if: runner.os == 'Linux'
-        run: ./test-builds.sh
+        env:
+          CC: ${{ matrix.compiler }}
+        run: |
+          test "x$CC" = "xclang" && export CXX=clang++
```

Comment:
> Please try to refactor to avoid as many of the following red flags as possible:
> 
> * Needlessly different commands for different OSes. I am aware that MacOS tests currently exclude gcc; ignore that ephemeral fact when refactoring.

There is little point in keeping them aligned, but sure.

> * CXX is only set for some compilers. I am aware that our ./configure does not need CXX to be set when CC is gcc; ignore that compiler-specific fact when refactoring.

It's not "our" configure. It's baked into any configure script.

> * An inconsistent mixture of `env FOO` and `export BAR` commands.

It's unavoidable, because "env:" commands do not support string manipulation. The next best option is to do it all in the shell script. Doesn't feel like an improvement to me, but ¯\_(ツ)_/¯

> * Avoidable conditionals

There is one unavoidable conditional


> One of the tools that may help you in this refactoring is this matrix trick:
> 
> ```yaml
> matrix:
>     os: [ ubuntu-22.04, macos-14 ]
>     compiler: [ {cxx: g++, cc: gcc}, {cxx: clang++, cc: clang} ]
> ```
> 
> or, if you prefer, this matrix trick:
> 
> ```yaml
> matrix:
>     os:
>         - ubuntu-22.04
>         - macos-14
> 
>     include:
>        - cxx_compiler: g++
>          c_compiler: gcc,
>        - cxx_compiler: clang++
>          c_compiler: clang,
> ```
> 
> I have neither [invented](https://stackoverflow.com/q/66025220) nor tested these tricks, but I do hope there is something better (like these tricks) than pairing compiler names using shell `test` statements!

We could use shell "if" statements and "case" statements as well :P

Interestingly, github's code editor complains about using a map for a matrix dimension (although it works). This could mean it might stop working in the future. Using it uglifies the exclusion list, but [it works](https://github.com/squid-cache/squid/actions/runs/9037244905/job/24835847230).

I'm following your suggestion, although unconvinced. Hopefully it's code we won't need to touch ever again